### PR TITLE
Adding gunzip filter.

### PR DESCRIPTION
**Pull request by peteclark-ft**
Despite saying it does, golang doesn't unzip gzipped requests. So we have to do that for it.

### CODE REVIEW

File: resources/filters.go [hunk 1]
```diff
@@ -29,3 +30,19 @@ func FilterSyntheticTransactions(next func(w http.ResponseWriter, r *http.Reques
 		next(w, r)
 	}
 }
+
+// UnzipGzip pre-processes the request body if it's gzipped.
+func UnzipGzip(next func(w http.ResponseWriter, r *http.Request)) func(w http.ResponseWriter, r *http.Request) {
```
Comment (izzyblues) [inline resources/filters.go]: Maybe the proper name should be "Gunzip"?

File: resources/filters_test.go [hunk 2]
```diff
@@ -54,3 +57,40 @@ func TestAllowsNoTID(t *testing.T) {
...
+	gz.Write([]byte("not gibberish"))
+
+	req, _ := http.NewRequest("PUT", "http://nothing/at/all", bytes.NewReader(b.Bytes()))
+	req.Header.Add("Content-Encoding", "gzip")
+
+	w := httptest.NewRecorder()
+	UnzipGzip(next)(w, req)
+
+	assert.Equal(t, 200, w.Code)
+	if !passed {
```
Comment (izzyblues) [inline resources/filters_test.go]: I think this statement should be `assert.True(t, passed, "Next handler should be reached")`

File: main.go [hunk 3]
```diff
@@ -107,7 +107,9 @@ func server(port int, maxSinceInterval int, dumpRequests bool, mapper mapping.No
 	r := mux.NewRouter()
 
 	r.HandleFunc("/lists/notifications", resources.ReadNotifications(mapper, nextLink, db, maxSinceInterval))
-	r.HandleFunc("/lists/{uuid}", resources.FilterSyntheticTransactions(resources.WriteNotification(dumpRequests, mapper, db))).Methods("PUT")
+
+	gunzipped := resources.UnzipGzip(resources.WriteNotification(dumpRequests, mapper, db))
+	r.HandleFunc("/lists/{uuid}", resources.FilterSyntheticTransactions(gunzipped)).Methods("PUT")
```
Comment (izzyblues) [inline main.go]: I think that what you tried to achieve here is to write a king of "fluent" API. 
These two lines are fine to me, but if you want to achieve the "nirvana" of fluent APIa, you may want to obtain something like this:
```
r.HandleFunc("/lists/{uuid}", resources.FilterSyntheticTransactions().UnzipGzip().WriteNotification(...).Methods("PUT")
```

File: resources/filters.go [hunk 4]
```diff
@@ -10,8 +11,36 @@ import (
...
+}
+
+// Filter creates a new composable filter.
+func Filter(next func(w http.ResponseWriter, r *http.Request)) Filters {
+	return Filters{next}
+}
+
 // FilterSyntheticTransactions will filter out incoming requests if they have a synthetic prefix.
-func FilterSyntheticTransactions(next func(w http.ResponseWriter, r *http.Request)) func(w http.ResponseWriter, r *http.Request) {
+func (f Filters) FilterSyntheticTransactions() Filters {
```
Comment (khatton-ft) [inline resources/filters.go]: Shouldn't these all be pointer receivers?

**Pull request merged**